paimon-core/src/main/java/org/apache/paimon/operation/PickFilesUtil.java

```diff
@@ -0,0 +1,184 @@
+                            .map(
+                                    f ->
+                                            pathFactory
+                                                    .createDataFilePathFactory(
+                                                            f.partition(), f.bucket())
+                                                    .toPath(f.fileName()))
+                            .collect(Collectors.toList());
+            Collections.reverse(dataFiles);
+            files.addAll(dataFiles);
+
+            // try to read index files
+            String indexManifest = snapshot.indexManifest();
+            if (indexManifest != null && indexFileHandler.existsManifest(indexManifest)) {
+                files.add(pathFactory.indexManifestFileFactory().toPath(indexManifest));
+
+                List<IndexManifestEntry> indexManifestEntries =
+                        retryReadingFiles(
+                                () -> indexFileHandler.readManifestWithIOException(indexManifest));
+                if (indexManifestEntries == null) {
+                    return Collections.emptyList();
+                }
+
+                indexManifestEntries.stream()
+                        .map(IndexManifestEntry::indexFile)
+                        .map(indexFileHandler::filePath)
+                        .forEach(files::add);
+            }
+
+            // add statistic file
+            if (snapshot.statistics() != null) {
+                files.add(pathFactory.statsFileFactory().toPath(snapshot.statistics()));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        return files;
+    }
+
+    private static void addManifestList(
+            List<Path> used, Snapshot snapshot, FileStorePathFactory pathFactory) {
+        used.add(pathFactory.toManifestListPath(snapshot.baseManifestList()));
+        used.add(pathFactory.toManifestListPath(snapshot.deltaManifestList()));
+        String changelogManifestList = snapshot.changelogManifestList();
+        if (changelogManifestList != null) {
+            used.add(pathFactory.toManifestListPath(changelogManifestList));
+        }
+    }
+
+    private static List<ManifestFileMeta> readAllManifestsWithIOException(
+            Snapshot snapshot, ManifestList manifestList) throws IOException {
+        List<ManifestFileMeta> result = new ArrayList<>();
+
+        result.addAll(manifestList.readWithIOException(snapshot.baseManifestList()));
+        result.addAll(manifestList.readWithIOException(snapshot.deltaManifestList()));
+
+        String changelogManifestList = snapshot.changelogManifestList();
+        if (changelogManifestList != null) {
+            result.addAll(manifestList.readWithIOException(changelogManifestList));
+        }
+
+        return result;
+    }
+
+    @Nullable
+    private static <T> T retryReadingFiles(OrphanFilesClean.ReaderWithIOException<T> reader)
```

Comment:
No need to use interface from `OrphanFilesClean`. This interface is simple, just create one in this file.

Reply:
done